File: 2025/java/senior/Senior5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Senior5 {
    private final static int MOD = 1000000 + 3;

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokens = new StringTokenizer(reader.readLine());

        int n = Integer.parseInt(tokens.nextToken());

        List<int[]> tasks = new ArrayList<>();
        long[][] tree = new long[1 << 21][2];
        for (int i = 0; i < n; i++) {
            long ans = tree[0][0];
            tokens = new StringTokenizer(reader.readLine());
            String op = tokens.nextToken();

            int pos = 0;
            if ("A".equals(op)) {
                int start = (int) ((Integer.parseInt(tokens.nextToken()) + ans - 1) % MOD);
                int duration = (int) ((Integer.parseInt(tokens.nextToken()) + ans) % MOD);
                pos = (1 << 20) + start;
                tasks.add(new int[]{start, duration});
                long[] treeNode = tree[pos];
                treeNode[1] += duration;
                treeNode[0] = start + treeNode[1];
            } else {
                int index = (int) ((Integer.parseInt(tokens.nextToken()) + ans) % MOD);
                int[] task = tasks.get(index - 1);
                pos = (1 << 20) + task[0];
                long[] treeNode = tree[pos];
                treeNode[1] -= task[1];
                if (treeNode[1] == 0) {
                    treeNode[0] = 0;
                } else {
                    treeNode[0] -= task[1];
                }
            }

            pos = (pos + 1) / 2 - 1;
            while (true) {
                long[] left = tree[pos * 2 + 1];
                long[] right = tree[pos * 2 + 2];
                tree[pos][0] = Math.max(left[0] + right[1], right[0]);
                tree[pos][1] = left[1] + right[1];

                if (pos == 0) {
                    break;
                }
                pos = (pos + 1) / 2 - 1;
            }

            System.out.println(tree[0][0]);
        }
    }
}

File: 2025/cpp/senior/senior5.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1000000 + 3;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  cin >> n;

  vector<pair<int, int>> tasks{};
  vector<pair<long, long>> tree((1 << 21) + 1, pair<long, long>(0L, 0L));
  for (int i = 0; i < n; i++) {
    long ans = tree[1].first;
    string op;
    cin >> op;

    int pos = 0;
    if (op == "A") {
      int s, t;
      cin >> s >> t;
      s = static_cast<int>((s + ans - 1) % MOD);
      t = static_cast<int>((t + ans) % MOD);
      pos = (1 << 20) + s + 1;
      tasks.emplace_back(s, t);
      auto node = tree[pos];
      tree[pos] = pair<long, long>(s + node.second + t, node.second + t);
    } else {
      int idx;
      cin >> idx;
      idx = static_cast<int>((idx + ans) % MOD);
      auto task = tasks[idx - 1];
      pos = (1 << 20) + task.first + 1;
      auto node = tree[pos];
      if (task.second == node.second) {
        tree[pos] = pair<long, long>(0, 0);
      } else {
        tree[pos] = pair<long, long>(node.first - task.second, node.second - task.second);
      }
    }

    pos = pos >> 1;
    while (true) {
      auto left = tree[pos << 1];
      auto right = tree[(pos << 1) + 1];
      tree[pos].first = max(left.first + right.second, right.first);
      tree[pos].second = left.second + right.second;

      if (pos == 1) {
        break;
      }
      pos = pos >> 1;
    }

    cout << tree[1].first << endl;
  }
}